overlord/hookstate/ctlcmd/health.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	shortHealthHelp = i18n.G("Report on snap's health")
+	longHealthHelp  = i18n.G(`
+
+The set-health command can signal to the system and the user that something is
+not well with the snap.
+
+Note the health is of the snap, not of the apps it contains; it’s up to the
+snap developer to determine how the health of the individual apps add up to
+the health of the snap as a whole.
+
+status can be one of
+
+- okay: the snap is healthy. This status takes no message and no code.
+
+- waiting: some resource the snap needs isn’t ready yet; the user just needs
+  to wait.  The message (and optionally the code) must explain what it’s
+  waiting for.
+
+- blocked: the user needs to do something for the snap to do something; the
+  message (and optionally the) must say what it is needs doing.
+
+- error: something is broken; the message (and optionally the code) must
+  explain what.
+
+`)
+)
+
+func init() {
+	addCommand("set-health", shortHealthHelp, longHealthHelp, func() command { return &healthCommand{} })
+}
+
+type healthPositional struct {
+	Status  string `positional-arg-name:"<status>" required:"yes" description:"a valid health status; required."`
+	Message string `positional-arg-name:"<message>" description:"a short human-readable explanation of the status (when not okay). Must be longer than 7 characters, and will be truncated if over 70. Message cannot be provided if status is okay, and is required otherwise."`
+}
+
+type healthCommand struct {
+	baseCommand
+	healthPositional `positional-args:"yes"`
+	Code             string `long:"code" value-name:"<code>" description:"optional tool-friendly value representing the problem that makes the snap unhealthy.  Not a number, but a word with 3-30 bytes matching [a-z](?:-?[a-z0-9]){2,}"`
+}
+
+var (
+	// "choice" on positional args doesn't work
+	validStatus = regexp.MustCompile(`^(?:okay|waiting|blocked|error)$`).MatchString
+	validCode   = regexp.MustCompile(`^[a-z](?:-?[a-z0-9]){2,}$`).MatchString
+)
+
+func (c *healthCommand) Execute([]string) error {
+	if c.Status == "okay" && (len(c.Message) > 0 || len(c.Code) > 0) {
+		return fmt.Errorf(`when status is "okay", message and code should be empty`)
+	}
+
+	status, err := healthstate.StatusLookup(c.Status)
+	if err != nil {
+		return err
+	}
+
+	if len(c.Code) > 0 {
+		if len(c.Code) < 3 || len(c.Code) > 30 {
+			return fmt.Errorf("error code should have between 3 and 30 bytes, got %d", len(c.Code))
+		}
+		if !validCode(c.Code) {
+			return fmt.Errorf("invalid error code %q", c.Code)
+		}
+	}
+
+	if status != healthstate.OkayStatus {
+		if len(c.Message) == 0 {
+			return fmt.Errorf(`when status is not "okay", message is required`)
+		}
+
+		// n is the number of runes
+		// j is the index of the 71st rune
+		var j, n int
+		for i, r := range c.Message {
+			if strutil.IsCtrl(r) {
+				return fmt.Errorf(`character %+q at position %d is not allowed`, r, i)
+			}
+			if n == 70 {
+				j = i
```

Comment:
Can we break here or am I misreading it?

Reply:
we could (although then we'd "accept" a message that contains bad chars as long as they are beyond a certain point in the message, which would make the issue harder to spot when developing). In any case I'm dropping this bit of validation in the next commit, and will add it back later -- keeps this pr minimal